File: itinerary-fe/src/app/utils/distanceUtils.ts
import { TouristAttraction } from "../models/touristAttraction";

export class DistanceUtils {
    static getDistanceFromLatLonInKm(touristAttractionA: TouristAttraction, touristAttractionB: TouristAttraction) {
        const lat1 = touristAttractionA.latitude;
        const lon1 = touristAttractionA.longitude;
    
        const lat2 = touristAttractionB.latitude;
        const lon2 = touristAttractionB.longitude;
    
        const R = 6371; // Radius of the earth in km
        const dLat = this.deg2rad(lat2-lat1);  // deg2rad below
        const dLon = this.deg2rad(lon2-lon1); 
        const a = 
          Math.sin(dLat/2) * Math.sin(dLat/2) +
          Math.cos(this.deg2rad(lat1)) * Math.cos(this.deg2rad(lat2)) * 
          Math.sin(dLon/2) * Math.sin(dLon/2)
          ; 
        const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a)); 
        const d = R * c; // Distance in km
        return d;
    }
      
    static deg2rad(deg: number) {
        return deg * (Math.PI/180)
    }
}

File: itinerary_server/src/main/java/com/dametto/itinerary_server/utils/DistanceUtils.java
package com.dametto.itinerary_server.utils;

import com.dametto.itinerary_server.jpa.TouristAttraction;

public class DistanceUtils {

    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    /*::  This function converts decimal degrees to radians             :*/
    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    private static double deg2rad(double deg) {
        return (deg * Math.PI / 180.0);
    }

    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    /*::  This function converts radians to decimal degrees             :*/
    /*:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::*/
    private static double rad2deg(double rad) {
        return (rad * 180.0 / Math.PI);
    }

    public static Double calculateDistance(TouristAttraction firstPoint, TouristAttraction secondPoint) {
        Double lon1 = firstPoint.getLongitude();
        Double lat1 = firstPoint.getLatitude();
        Double lon2 = secondPoint.getLongitude();
        Double lat2 = secondPoint.getLatitude();

        double theta = lon1 - lon2;
        double dist = Math.sin(deg2rad(lat1)) * Math.sin(deg2rad(lat2)) + Math.cos(deg2rad(lat1)) * Math.cos(deg2rad(lat2)) * Math.cos(deg2rad(theta));
        dist = Math.acos(dist);
        dist = rad2deg(dist);
        dist = dist * 60 * 1.1515;
        dist = dist * 1.609344; // to KM
        return (dist);
    }
}
